onnxruntime/core/providers/cuda/tensor/tile_impl.cu

```diff
@@ -7,140 +7,178 @@
 namespace onnxruntime {
 namespace cuda {
 
-constexpr int MAX_DIMS = 16;
+#ifdef USE_ROCM
+constexpr int num_elements_per_thread = 2;
+constexpr int num_threads_per_block = 512;
+#else
+constexpr int num_elements_per_thread = GridDim::maxElementsPerThread;
+constexpr int num_threads_per_block = GridDim::maxThreadsPerBlock;
+#endif
 
 template <typename T>
-__global__ void _UnRolledTileKernel(
-    const size_t shape_rank,
-    const TArray<fast_divmod> fdm_input_shape,
-    const TArray<int64_t> input_strides,
-    const T* input_data,
-    const TArray<fast_divmod> fdm_output_strides,
-    T* output_data,
-    const CUDA_LONG N) {
-  CALCULATE_ELEMENTWISE_INDEX_OR_EXIT(id, N);
-  CUDA_LONG input_index = 0;
-  CUDA_LONG output_index = id;
+__global__ void _UnRolledTileKernel(const size_t shape_rank, const TArray<fast_divmod> fdm_input_shape,
+                                    const TArray<int64_t> input_strides, const T* input_data,
+                                    const TArray<fast_divmod> fdm_output_strides, T* output_data, const CUDA_LONG N) {
+  CUDA_LONG start = num_elements_per_thread * num_threads_per_block * blockIdx.x + threadIdx.x;
+  T value[num_elements_per_thread];
+  CUDA_LONG id = start;
+#pragma unroll
+  for (int i = 0; i < num_elements_per_thread; ++i) {
+    if (id < N) {
+      CUDA_LONG input_index = 0;
+      CUDA_LONG offset = id;
+#pragma unroll
+      for (auto dim = 0; dim < fdm_output_strides.Capacity(); ++dim) {
```

Comment:
When you initialize a TArray and put data that size are > 8, we will need to use a capacity number that > 8 other than the default value (for examle. TArray<fast_divmod, 16>), otherwise, you will not able to store your data to such TArray. I.E., if fdm_output_strides here is a valid TArray, and assume its size is larger than 8, then the capacity must be a number that larger than 8.

Reply:
I mean in Tile compute code, when we construct such a TArray, we did not check whether 8 is enough for hold the real rank.

 TArray<fast_divmod> fdm_output_strides(rank);
  TensorPitches output_pitches(output_dims);
  for (auto i = 0; i < rank; i++) {
    fdm_output_strides[i] = fast_divmod(static_cast<int>(output_pitches[i]));
  } 

But seems other TArray usage also did not do the check, so it is fine.

Reply:
You are right, these are all bugs, or they are based on some assumptions.